### PR TITLE
lib/upgrade: Enforce limits on download archives (fixes #3045)

### LINKED ISSUE #3045: A specially crafted download package can cause DoS
By creating an infinitely large temp file for the binary, for example. As reported privately by Sebastian Py. Thanks Sebastian!

### PR DESCRIPTION
**Pull request by calmh**
### Purpose

Enforce reasonable limits on sizes of downloaded archives and created files.

### CODE REVIEW

File: lib/upgrade/upgrade_supported.go [hunk 1]
```diff
@@ -25,19 +25,46 @@ import (
...
+const (
+	// Current binary size hovers around 10 MB. We give it some room to grow
+	// and say that we never expect the binary to be larger than 64 MB.
+	maxBinarySize = 64 << 20 // 64 MiB
+
+	// The max expected size of the signature file.
+	maxSignatureSize = 1 << 10 // 1 KiB
+
+	// We set the same limit on the archive. The binary will compress and we
+	// include som other stuff - currently the release archive size is
```
Comment (AudriusButkevicius) [inline lib/upgrade/upgrade_supported.go]: Some

**Comment by AudriusButkevicius** [comment]
@st-review merge

**Comment by st-review** [comment]
:ok_hand: Merged as 38166e976f0b5e58c1192d6d070cf769a4193b0c. Thanks, @calmh!

**Comment by AudriusButkevicius** [comment]
Does this address the bombs?

**Comment by AudriusButkevicius** [comment]
I guess, the question is, did you try to give it a bomb archive to see if this prevents this?

**Comment by calmh** [comment]
I tried it with an archive with a 10G `syncthing` in it, which it aborts after a few seconds:

```
jb@syno:~/s/g/s/syncthing $ STTRACE=upgrade ./bin/syncthing -upgrade-to https://nym.se/t/syncthing-linux-amd64-v0.99.0.tar.gz
2016/05/06 16:05:31.518732 upgrade_supported.go:178: DEBUG: loading "https://nym.se/t/syncthing-linux-amd64-v0.99.0.tar.gz"
2016/05/06 16:05:32.067221 upgrade_supported.go:296: DEBUG: considering file syncthing
2016/05/06 16:05:32.067259 upgrade_supported.go:305: DEBUG: found upgrade binary syncthing
2016/05/06 16:05:40.201561 main.go:339: FATAL: Upgrade: no signature found
```

There is one more thing that just hit me that should probably be added...

**Comment by AudriusButkevicius** [comment]
Yeah but what about a 1kb archive which explodes into GB's?

**Pull request closed without merging**